### DISCUSSION TITLE
How do I render custom menu using the INavigationManager?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#8238] [Q&A]
I created user profile page in my app. Now, I need to add custom menu to the profile so the user can navigate between the tabs.

I am using a wrapper to allow me to render all shapes of the profile pages. In the wrapper, I am injecting an instance of the `INavigationManager` to build my custom menu using the `ClientProfile` name.

```
@inject OrchardCore.Navigation.INavigationManager navigationManager;

<div class="card text-center">
    <div class="card-header pb-0">
        @if (Model.Header != null)
        {
            @await DisplayAsync(Model.Header)
        }
        @await DisplayAsync(await navigationManager.BuildMenuAsync("ClientProfile", Html.ViewContext))       
    </div>

    @if (Model.TabContent != null)
    {
        <div class="card-body">
            @await DisplayAsync(Model.TabContent)
        </div>
    }
</div>
```

But the `@await DisplayAsync(await navigationManager.BuildMenuAsync("ClientProfile", Html.ViewContext))` line is displaying the following text not the actual menu. 

 > System.Collections.Generic.List`1[OrchardCore.Navigation.MenuItem] ` 

I am guessing I need to create views for the `MenuItem` so `OC` knows how to render the menu as tabs. But I am not sure what should the view be called here so I can create it? Also, is `@await DisplayAsync(await navigationManager.BuildMenuAsync("ClientProfile", Html.ViewContext))` the correct way to render the menu?

**Comment by jtkech** [comment]
@CrestApps 

Here i think you want to use the local navigation with TABs that can be defined through `Location` in drivers / placement info, placement info can be defined in a `Placement.json` file or through the admin UI with the placement feature enabled (see at the end)

Here, some info to follow how it works

In `ZoneShapes`

    Shape container = (Shape)await ShapeFactory.CreateAsync("TabContainer", Arguments.From(

In `TabContainer.cshtml`

    @await DisplayAsync(await New.LocalNavigation(ContentItem: Model.ContentItem, Tabs: Model.Grouping))

In `LocalNavigation.cshtml`

    @await DisplayAsync(await New.AdminTabs(ContentItem: Model.ContentItem, Tabs: Model.Tabs))

Then see what `AdminTabs.cshtml` does in this loop

    @foreach (var tab in tabs)

For my tests i defined the 2 following placements

TextField

    [
      {
        "place": "Header:0#Text;0",
        "displayType": "Detail",
        "differentiator": null,
        "alternates": null,
        "wrappers": null,
        "shape": null,
        "contentType": [
          "Article"
        ]
      }
    ]

    MediaField

    [
      {
        "place": "Header:0#Media;0",
        "displayType": "Detail",
        "differentiator": null,
        "alternates": null,
        "wrappers": null,
        "shape": null,
        "contentType": [
          "Article"
        ]
      }
    ]

Here the result

![NavBar](https://user-images.githubusercontent.com/8586360/103605757-044d1480-4f15-11eb-9bfc-e8e3fc9faf9e.png)

**Reply by MikeAlhayek** [reply]
@jtkech is there an example somewhere that I can follow? I need to create the tabs from code and from multiple modules. I though using INavigationProvider would be best/easiest way since any module can define them.

**Reply by jtkech** [reply]
Both are possible but the INavigationProvider is more suitable to generate links

With tabs each module can have a driver providing a shape with a given `Location` defining the tab name / order to use, and that can be overridden by a placement defined through a file or through the admin UI

See https://docs.orchardcore.net/en/dev/docs/reference/core/Placement/#supported-groupings
And https://docs.orchardcore.net/en/dev/docs/reference/modules/Placements/

To see how to define the Location string

**Reply by MikeAlhayek** [reply]
@jtkech that sounds good. Tabs do seems to also support tab with drop down like something like this

![image](https://user-images.githubusercontent.com/24724371/103670222-243d0200-4f2e-11eb-8744-9f97d6ca77c6.png)

So I would need to change my profile wrapper so this, right?

```
<div class="card text-center">
    <div class="card-header pb-0">
        @if (Model.Header != null)
        {
            @await DisplayAsync(Model.Header)
        }
        @if(Model.TabContainer != null) 
       {
            @await DisplayAsync(Model.TabContainer )
       }
    </div>

    @if (Model.TabContent != null)
    {
        <div class="card-body">
            @await DisplayAsync(Model.TabContent)
        </div>
    }
</div>
```

But how will my driver look like that generates the tabs and sub-tabs? I mean what ViewModel would I initialize to provide the links to the tabs. My tabs point to a a controller, area when then are clicked.

**Reply by MikeAlhayek** [reply]
@jtkech  Actually looking the the `AdminTabs.cshtml` It does not look like it supports links. Like when a user on clicks on the overview tab, take them to `/profile/overview/{id}` and when the user clicks on the Biometrics tab redirect them to `/biometrics/index/{id}`. Also, I don't think there is support for dropdown buttons. I could be wrong here but I don't see how the view support this.

**Reply by MikeAlhayek** [reply]
@jtkech since I am not convince that tabs will do what I am looking for, I am back exploring rendering menu using `INavigationManager`. If you think that I can create table, with links to other pages and sub menu in a tab, please advise.

In an attempt to render the menu items, I changed my wrapper to the following

```
@using OrchardCore.Navigation;
@inject OrchardCore.Navigation.INavigationManager NavigationManager;
@inject OrchardCore.DisplayManagement.IDisplayManager<MenuItem> MenuItemDisplayManager;
@inject OrchardCore.DisplayManagement.ModelBinding.IUpdateModelAccessor ModelUpdaterAccessor;

<div class="card text-center">
    <div class="card-header pb-0">

        @if (Model.Header != null)
        {
            @await DisplayAsync(Model.Header)
        }
        @{
            var menuItems = await NavigationManager.BuildMenuAsync("ClientProfile", Html.ViewContext);
            var updater = ModelUpdaterAccessor.ModelUpdater;
            <ul class="nav nav-tabs card-header-tabs">

                @foreach (var menuItem in menuItems.Where(x => x != null))
                {
                    dynamic node = await MenuItemDisplayManager.BuildDisplayAsync(menuItem, updater, "ClientProfile");
                    @await DisplayAsync(node)
                }
            </ul>
        }

    </div>

    @if (Model.TabContent != null)
    {
        <div class="card-body">
            @await DisplayAsync(Model.TabContent)
        </div>
    }
</div>
```

Then I created `MenuItem.ClientProfile.cshtml` view to render the menuItem you see below. But running into an issue where `Model.MenuItem` is always null. How can I access the MenuItem from the model so I can render it?

```
@model dynamic
@using OrchardCore.Navigation;

@{
    MenuItem menuItem = Model.MenuItem;

    if(menuItem == null)
    {
        return;
    }
    var children = menuItem.Items;
}

@if (children != null && children.Any())
{
    <li class="nav-item dropdown">
        <a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#" role="button" aria-haspopup="true" aria-expanded="false">@menuItem.Text</a>
        <div class="dropdown-menu">
            @foreach (var child in children)
            {
                <a class="dropdown-item" href="@child.Href">@child.Text</a>
            }
        </div>
    </li>
}
else
{
    <li class="nav-item">
        <a class="nav-link" href="@menuItem.Href">@menuItem.Text</a>
    </li>
}
```

**Comment by jtkech** [comment]
@CrestApps 

Okay, didn't think you need links, so yes, as said `INavigationProvider` is more suitable to generate links

Normally in your view the model is your menuitem, hmm but looking at the source code, try both

                    @await DisplayAsync(node.Content)
                    @await DisplayAsync(node)

And in your view try both, and still use it as a dynamic object

    var menuItem = Model;
    var menuItem = Model.Content;

Or try to use a `@model` directive on `MenuItem` as it may have built a proxy on this type.

Hmm, you may have to follow what `NavigationShapes` does, it calls `BuildMenuAsync()` => `PopulateMenuAsync()` => ....., => `BuildMenuItemShapeAsync()` that uses a `shapeFactory` to build `menuItemShape`

**Update**: Also try to put a break point in your view to check the underlying type and what it contains

**Reply by MikeAlhayek** [reply]
@jtkech I added a break point into my view and the model type is `OrchardCore.DisplayManagement.Zones.ZoneHolding`
![image](https://user-images.githubusercontent.com/24724371/103780953-8bb88780-4fea-11eb-9c55-2c6d2476402a.png)

Also, changing `@await DisplayAsync(node)` to `@await DisplayAsync(node.Content)` does nothing since `Content` shape does not exists.

I tried calling `PopulateMenuAsync` like this 

```
<ul class="nav nav-tabs card-header-tabs">

    @foreach (var menuItem in menuItems.Where(x => x != null))
    {
        dynamic node = await MenuItemDisplayManager.BuildDisplayAsync(menuItem, updater, "ClientProfile");
        await NavigationHelper.PopulateMenuAsync(New, node, node, menuItems, Html.ViewContext);
        @await DisplayAsync(node)
    }
</ul>
```
which gave me the following error

 > RuntimeBinderException: The best overloaded method match for 'Microsoft.AspNetCore.WebUtilities.QueryHelpers.AddQueryString(string, string, string)' has some invalid arguments

Looking at the [MenuItem.TreeSummary.cshtml](https://github.com/OrchardCMS/OrchardCore/blob/87b5619a8a4e1a1a1e5e854d323b9191a4c052f8/src/OrchardCore.Modules/OrchardCore.AdminMenu/Views/MenuItem.TreeSummary.cshtml)  which is accessing the `Model.MenuItem;` like I am doing. But in my case `MenuItem` does not exists

**Reply by MikeAlhayek** [reply]
I ended up using standard Razor views instead of shapes to render the menu items.

**Reply by jtkech** [reply]
@CrestApps 

Just for info, in debug mode when you are looking at a shape / model (as you did above and where it is a `ZoneHolding` that is a shape), to get a deeper look of the shape / model, expand the `Raw View`, you will see e.g. a `Properties` field that may contain other objects, e.g. child shapes but not only, also look at non public members as e.g. `_items` that may contain child shapes, this because the `Items` property only say that it is an `IEnumerable<object>`
